Pin images by digest

nishanttotla opened the pull request:
**- What I did**
When a Docker Swarm mode service is created using the image and tag, the agents pull image with the specified tag. This creates an issue for non-static tags like `latest`, which keep updating. This might cause multiple versions of the image running for the same service. The solution to this problem is to pin images by digest.

For example, on `docker service create alpine:latest top`, the manifest for the `alpine:latest` image is retrieved, and the digest is added to the image description before passing on the spec to the SwarmKit manager. In this case, `alpine:latest` is changed to `alpine:latest@sha256:1354db23ff5478120c980eca1611a51c9f2b88b61f24283ee8200bf9a54f2e5c` that specifies an image precisely. If at a future time, the image under `alpine:latest` were to have a different digest, this would be detected at the next `service update`, and all images would be pinned to the new digest.

Related issues: #24066, #27508, https://github.com/docker/swarmkit/issues/1334

**- How I did it**
The change is fairly simple. A new daemon function `ResolveTagToDigest` is used to retrieve a digest for a `reference.NamedTagged` object, by making a request to the registry. This function is used by `docker service create` and `docker service update` to resolve image to digest, if the image used to create the service didn't already have it.

The service spec is modified before passing to the SwarmKit manager to always contain a digest string, thereby ensuring that a service image is always precisely specified. On image update, a digest is retrieved again, and if different from the existing one, will result in a service update.

One thing that remains to be done is to enable pinning by digest for private images (this is a minor addition which I will shortly finish) by passing `AuthConfig` to `ResolveTagToDigest`.

**- How to verify it**
Try the following commands:
```
docker service create --name testsvc alpine:latest top
docker service ls
```
This should show the image as `alpine:latest@sha256:1354db23ff5478120c980eca1611a51c9f2b88b61f24283ee8200bf9a54f2e5c`
```
docker service update --image alpine:edge testsvc
docker service ls
```
This will show the image updated to `alpine:edge@sha256:cd9c03c2d382fcf00c31dc1635445163ec185dfffb51242d9e097892b3b0d5b4`

NOTE: It is by choice that the operation won't fail if the digest retrieval fails for some reason. It will only print a warning and move on. This is to make sure that the use case where a locally built image might be used to create a service still works.

**- Description for the changelog**
<!--
Write a short (one line) summary that describes the changes in this
pull request for inclusion in the changelog:
-->
Pin images by digest for docker service create and update

**- A picture of a cute animal (not mandatory but encouraged)**

<img width="743" alt="screen shot 2016-11-08 at 10 07 29 am" src="https://cloud.githubusercontent.com/assets/1872537/20111030/9a1fba9e-a59b-11e6-9eb5-4a167e5e476b.png">

This is Mutsu the kitten. For more of Mutsu, you can check [here](https://www.youtube.com/watch?v=PJpcEuSrEn8)

nishanttotla committed:
Adding function to resolve tag to digest in daemon

Signed-off-by: Nishant Totla <nishanttotla@gmail.com>

nishanttotla committed:
Pin image by digest on service create and update

Signed-off-by: Nishant Totla <nishanttotla@gmail.com>

aaronlehmann commented on daemon/cluster/cluster.go:
```diff
@@ -871,6 +872,34 @@ func (c *Cluster) GetServices(options apitypes.ServiceListOptions) ([]types.Serv
+			return "", err
+		}
+		logrus.Infof("pinned image to digest: %s", dgst)
+		return image + "@" + dgst, nil
```
`return reference.WithDigest(ref, dgst).String()`

nishanttotla replied:
Fixed.

nishanttotla commented:
cc @aaronlehmann @aluzzardi @tiborvass @icecrime @thaJeztah

aaronlehmann commented on daemon/image_pull.go:
```diff
@@ -104,3 +104,41 @@ func (daemon *Daemon) pullImageWithReference(ctx context.Context, ref reference.
+	digest := ""
+
+	// only retrieve digest if the repo is v2
+	if confirmedV2 {
```
I'd suggest ignoring `confirmedV2` here and attempting the `Tags(ctx).Get()` anyway. Worst case, it fails.

nishanttotla replied:
Sounds fair. The check doesn't provide any particular advantages anyway.

aaronlehmann commented on daemon/image_pull.go:
```diff
@@ -104,3 +104,41 @@ func (daemon *Daemon) pullImageWithReference(ctx context.Context, ref reference.
+		if err != nil {
+			return "", err
+		} else {
+			digest = dscrptr.Digest.String()
```
Return the `Digest` object rather than the stringified version.

aaronlehmann commented on daemon/cluster/cluster.go:
```diff
@@ -871,6 +872,34 @@ func (c *Cluster) GetServices(options apitypes.ServiceListOptions) ([]types.Serv
+		// add default tag if not present
+		if reference.IsNameOnly(ref) {
+			ref = reference.WithDefaultTag(ref)
+		}
```
The `if` is not necessary - you can just do `ref = reference.WithDefaultTag(ref)` since `WithDefaultTag` is defined as follows:

```go
// WithDefaultTag adds a default tag to a reference if it only has a repo name.
func WithDefaultTag(ref Named) Named {
        if IsNameOnly(ref) {
                ref, _ = WithTag(ref, DefaultTag)
        }
        return ref
}
```

nishanttotla replied:
You're right. I think I missed this after some refactoring.

aaronlehmann commented on daemon/cluster/cluster.go:
```diff
@@ -871,6 +872,34 @@ func (c *Cluster) GetServices(options apitypes.ServiceListOptions) ([]types.Serv
+		if reference.IsNameOnly(ref) {
+			ref = reference.WithDefaultTag(ref)
+		}
+		namedTaggedRef, _ := ref.(reference.NamedTagged)
```
namedTaggedRef, ok := ref.(reference.NamedTagged)

Return an error if `ok` is false (should not be possible currently, but the reference package could change in the future).

aaronlehmann commented on daemon/cluster/cluster.go:
```diff
@@ -871,6 +872,34 @@ func (c *Cluster) GetServices(options apitypes.ServiceListOptions) ([]types.Serv
+			logrus.Warnf("unable to retrieve digest for image: %s", err.Error())
+			return "", err
+		}
+		logrus.Infof("pinned image to digest: %s", dgst)
```
This should probably be a `Debugf`, and ideally should provide more context about what was pinned. I'd suggest creating a logger at the top of this function with `logrus.WithFields`. Then if you use that logger for the log messages, they can have the service name, image reference, and so on filled in.

nishanttotla replied:
I moved these log messages outside this function, which now only returns an error, and the caller decides what to do with it.

aaronlehmann commented on daemon/cluster/cluster.go:
```diff
@@ -980,6 +1016,21 @@ func (c *Cluster) UpdateService(serviceIDOrName string, version uint64, spec typ
+	ctnr := serviceSpec.Task.GetContainer()
+	if ctnr == nil {
+		return fmt.Errorf("service does not use container tasks")
+	}
```
Move this above the `if encodedAuth != "" {` so `cntr` can be reused there.

nishanttotla commented:
@aaronlehmann I've addressed all your comments. Please let me know if this looks acceptable. The only thing I haven't quite decided is what fields to add to `logrus.Debugf`, though I moved the log statements to the caller so that all logging related to digest pinning is in one place.

Also cc @tonistiigi.

aaronlehmann commented on daemon/cluster/cluster.go:
```diff
@@ -979,7 +1021,20 @@ func (c *Cluster) UpdateService(serviceIDOrName string, version uint64, spec typ
+		logrus.Debugf("pinning image %s by digest: %s", newCtnr.Image, digestImage)
+		newCtnr.Image = digestImage
+	} else {
+		logrus.Debugf("unable to pin image %s to digest: %s", newCtnr.Image, err.Error())
```
This would log an error message if the user already provided a digest. How about making `imageWithDigestString` return `image, nil` if the image already has a digest specified, and then this code can only log "pinning image..." if `digestImage != newCntr.Image`?

stevvooe commented on daemon/cluster/cluster.go:
```diff
@@ -871,6 +872,36 @@ func (c *Cluster) GetServices(options apitypes.ServiceListOptions) ([]types.Serv
+		}
+		return refDgst.String(), nil
+	}
+	return "", fmt.Errorf("error while trying to retrieve image digest")
```
Don't make the error path in the main path. Get in the habit of making an early return.

stevvooe commented on daemon/cluster/cluster.go:
```diff
@@ -871,6 +872,36 @@ func (c *Cluster) GetServices(options apitypes.ServiceListOptions) ([]types.Serv
+	if err != nil {
+		return "", err
+	}
+	// only pin if not a canonical reference (i.e. with digest)
```
Why wouldn't we return the canonical reference if it is already canonical?

```
if canonical, ok := ref.(reference.Canonical); ok {
  return canonical, nil
}
```

stevvooe commented on daemon/cluster/cluster.go:
```diff
@@ -871,6 +872,36 @@ func (c *Cluster) GetServices(options apitypes.ServiceListOptions) ([]types.Serv
 
+// imageWithDigestString takes an image such as name or name:tag
+// and returns the image pinned to a digest, such as name@sha256:34234...
+func (c *Cluster) imageWithDigestString(ctx context.Context, image, encodedAuth string) (string, error) {
```
`encodedAuth` should really be cross-cutting. Perhaps, store in the backed or preconfigure an object with the auth before calling this method.

nishanttotla replied:
@stevvooe if I understand correctly, passing the `types.AuthConfig` struct to this function would be acceptable?

stevvooe replied:
That is probably a first step. It is hard to tell, since it is not actually being used. I have no clue why we would bury an unencoded value so deeply in the call stack. The auth should probably be decoded as soon as the request is received. Then, I would put it in the context and access as needed, without passing it down the entire call chain.

stevvooe commented on daemon/image_pull.go:
```diff
@@ -104,3 +104,39 @@ func (daemon *Daemon) pullImageWithReference(ctx context.Context, ref reference.
 	return err
 }
+
+func (daemon *Daemon) ResolveTagToDigest(ctx context.Context, ref reference.NamedTagged, authConfig *types.AuthConfig) (digest.Digest, error) {
```
Rather than add yet another method to daemon, why not export access to the configured repository? As it is, this method is very narrow and not very flexible. Ideally, we give it an image reference and get back fully configured repository.

nishanttotla replied:
Will make this function return a fully configured repository to expand its usability.

aluzzardi commented on daemon/cluster/cluster.go:
```diff
@@ -893,14 +924,24 @@ func (c *Cluster) CreateService(s types.ServiceSpec, encodedAuth string) (string
 
+	// pin image by digest
+	digestImage, err := c.imageWithDigestString(ctx, ctnr.Image, encodedAuth)
+	if err == nil {
```
if err != nil {
log warning
} else if digestImage != ctnr.Image {
log debug(pinning)
ctnr.Image = digestImage
}

aluzzardi reviewed (commented):
Looks good overall - just missing auth

tonistiigi commented:
@nishanttotla Return early for `DOCKER_SERVICE_PREFER_OFFLINE_IMAGE` to unlock the tests. https://github.com/docker/docker/blob/master/daemon/cluster/executor/container/controller.go#L91

nishanttotla commented:
I've addressed all comments, waiting on tests to pass now.

Still thinking about the right way to do auth, but the patch works (it will only pin to digest for public images). If it makes sense, I could also follow up with the auth stuff in a separate PR after the code freeze. 

cc @aaronlehmann @tiborvass @tonistiigi

aaronlehmann commented:
@nishanttotla: It sound be straightforward to convert `encodedAuth` to an `AuthConfig` struct. Here is the code from `daemon/cluster/executor/container/adapter.go`:

```go
        authConfig := &types.AuthConfig{}
        if encodedAuthConfig != "" {
                if err := json.NewDecoder(base64.NewDecoder(base64.URLEncoding, strings.NewReader(encodedAuthConfig))).Decode(authConfig); err != nil {
                        logrus.Warnf("invalid authconfig: %v", err)
                }
        }
```

Possibly we should factor this into a utility function (though I'm not sure what the best place is). This could be done in a followup.

nishanttotla commented:
@aaronlehmann thanks. Looks like this works. 

As a side note, I also find it unpleasant to decode the auth here (or in general at any point along the path), but don't see any better way to make it cleaner right now.

@tonistiigi all tests but one pass now. Not sure if this one is related to my change, I'm looking into it:
```
02:44:57 FAIL: docker_cli_swarm_test.go:952: DockerSwarmSuite.TestSwarmRotateUnlockKey
02:44:57 
02:44:57 [d761dbbbf650b] waiting for daemon to start
02:44:57 [d761dbbbf650b] daemon started
02:44:57 [d761dbbbf650b] exiting daemon
02:44:57 [d761dbbbf650b] waiting for daemon to start
02:44:57 [d761dbbbf650b] daemon started
02:44:57 [d761dbbbf650b] exiting daemon
02:44:57 [d761dbbbf650b] waiting for daemon to start
02:44:57 [d761dbbbf650b] daemon started
02:44:57 docker_cli_swarm_test.go:1020:
02:44:57     c.Assert(err, checker.IsNil, check.Commentf("out: %v", string(out)))
02:44:57 ... value *exec.ExitError = &exec.ExitError{ProcessState:(*os.ProcessState)(0xc421b583c0), Stderr:[]uint8(nil)} ("exit status 1")
02:44:57 ... out: Error response from daemon: swarm could not be unlocked: invalid key provided
02:44:57 
02:44:57 
02:44:57 [d761dbbbf650b] exiting daemon
02:44:58 
02:44:58 ----------------------------------------------------------------------
```

aaronlehmann commented on daemon/image_pull.go:
```diff
@@ -104,3 +105,25 @@ func (daemon *Daemon) pullImageWithReference(ctx context.Context, ref reference.
+	}
+
+	// retrieve repository
+	// TODO(nishanttotla): More sophisticated selection of endpoint
```
Yes, this should try all endpoints with `Version == registry.APIVersion2` in order until one succeeds.

nishanttotla replied:
@aaronlehmann is it reasonable to copy the code in `distribution/pull.go`? Looks like a bunch of reproduced code, which could also be a candidate for refactoring later.

nishanttotla replied:
@aaronlehmann: fixed, now endpoints are searched in order until one succeeds.

aaronlehmann commented on daemon/cluster/cluster.go:
```diff
@@ -893,14 +934,27 @@ func (c *Cluster) CreateService(s types.ServiceSpec, encodedAuth string) (string
+			logrus.Debugf("pinning image %s by digest: %s", ctnr.Image, digestImage)
+			ctnr.Image = digestImage
+		}
+	}
```
Seems like this code should be its own function, as the same code is added below in `UpdateService`.

nishanttotla replied:
There's a slight difference between the two, since in `UpdateService`, we also check if the image was updated, and print a Debug message.

aaronlehmann commented:
That test failure is probably not related. The test was just added in #27967. I'm going to look into it.

nishanttotla commented:
Test failure seems unrelated. I've also tested the case of image digest updating under the same tag, with and without registry auth. Is this ready to merge? I can refactor code and submit follow up PRs after the code freeze.
 @aaronlehmann @tonistiigi @aluzzardi

tiborvass commented on daemon/cluster/cluster.go:
```diff
@@ -893,14 +935,33 @@ func (c *Cluster) CreateService(s types.ServiceSpec, encodedAuth string) (string
+	authConfig := &apitypes.AuthConfig{}
+	if encodedAuth != "" {
+		if err := json.NewDecoder(base64.NewDecoder(base64.URLEncoding, strings.NewReader(encodedAuth))).Decode(authConfig); err != nil {
+			logrus.Warnf("invalid authconfig: %v", err)
```
is this expected not to error out?

nishanttotla replied:
Any operations related to pinning by digest aren't supposed to error out right now, because some people use local images in dev, and this would break that workflow.

nishanttotla commented on daemon/cluster/cluster.go:
```diff
@@ -871,6 +873,46 @@ func (c *Cluster) GetServices(options apitypes.ServiceListOptions) ([]types.Serv
+		// }
+		// return namedDigestedRef.String(), nil
+		return image + "@" + dscrptr.Digest.String(), nil
+	} else {
```
The `else` here makes sure that if the image provided already contained a digest, then this function returns it. This makes it simpler for the caller, which can assume that it always gets a string with digest returned to it.

aaronlehmann commented:
One thing I just thought of is that we should avoid re-pulling the image when the task starts if the image is pinned by digest and it already exists locally. This will make task startup faster and reduce unnecessary registry traffic.

This should be checked in the `Prepare` function in `daemon/cluster/executor/container/controller.go`. I think we may need to add `LookupImage` to `Backend` in `daemon/cluster/executor/backend.go` and use that function to check if the image already exists locally.

I'd suggest doing this in a separate PR before the release.

nishanttotla commented:
@aaronlehmann I will open a docker issue tracking these TODO items then.
- Revendor `distibution` to use the updated `WithDigest` from https://github.com/docker/distribution/pull/2044
- Change `service ls/ps/inspect` UI to suppress tag
- Avoid pulling image from registry if it exists locally
- Further refactoring (wherever necessary)

aaronlehmann commented:
Tested and this works for me.

LGTM

Thanks for that list of action items. It looks good to me. I would add one more - add an integration test that covers pinningby digest.

tonistiigi commented:
LGTM

aluzzardi commented:
LGTM

nishanttotla commented:
@thaJeztah this needs a `docs/revisit` label, I'm working on updating docs for this.
